api/envoy/config/cluster/aggregate/v2alpha/cluster.proto

```diff
@@ -0,0 +1,20 @@
+syntax = "proto3";
+
+package envoy.config.cluster.aggregate.v2alpha;
+
+option java_outer_classname = "ClusterProto";
+option java_multiple_files = true;
+option java_package = "io.envoyproxy.envoy.config.cluster.aggregate.v2alpha";
+
+import "validate/validate.proto";
+
+// [#protodoc-title: Aggregate cluster configuration]
+
+// Configuration for the aggregate cluster. See the :ref:`architecture overview
+// <arch_overview_aggregate_cluster>` for more information.
+// [#extension: envoy.clusters.aggregate]
+message ClusterConfig {
+  // Load balancing clusters in aggregate cluster. Clusters are prioritized based on the order they
```

Comment:
I suspect at some point in the future there will be demand for not having this list ordering. Consider your examples from the docs:

> from cluster with 0.1s connection timeout to cluster with 1s connection timeout, etc.

Why does this require failover? It seems perfectly valid to have some endpoints with 0.1s timeout and others with 1s timeout and round robin between these. By having the list ordering we are cutting off these use cases.

I also think it causes some confusion on when I should use locality priority and when I should use aggregate cluster. Should we just throw out the locality priority stuff and send a cluster per priority instead?

Anyhow I don't want to push back on this, but is it possible to craft the API in a way such that this sort of expansion in the future is possible? Possible a list of `(cluster, priority)` pair, and for now we can enforce priorities are unique, then in the future we could possibly remove this restriction without needing a major api change.

Reply:
> Why does this require failover? It seems perfectly valid to have some endpoints with 0.1s timeout and others with 1s timeout and round robin between these. By having the list ordering we are cutting off these use cases.

Selecting the host at the aggregate cluster level doesn't make sense because the aggregate cluster doesn't know the load balancing policy of the including clusters. The prerequisite is that both clusters use round robin. In this case, aggregate cluster can merge them together and do load balancing at the aggregate cluster level but it will make aggregate cluster more complex during load balancing. 

> I also think it causes some confusion on when I should use locality priority and when I should use aggregate cluster. Should we just throw out the locality priority stuff and send a cluster per priority instead?

I don't think that's conflict. The current locality weight load balancing has three steps: 
```
  1. Pick priority level. 
  2. Pick locality (as described in this section) within priority level from (1).
  3. Pick endpoint using cluster specified load balancer within locality from (2).
```

It allows such scenario:
```
P0 us_east{H0}
P1 us_east {H1, H2}, us_west {H3}
```

The load balancer could select H0 even though the locality is in us_west because the first step in locality weighted load balancing is selecting priority. This never happens for istio because istio always set same priority for endpoints.

By adding aggregate cluster, we introduced a new step, i.e., pick the cluster which happens before picking locality. 

> Anyhow I don't want to push back on this, but is it possible to craft the API in a way such that this sort of expansion in the future is possible? Possible a list of (cluster, priority) pair, and for now we can enforce priorities are unique, then in the future we could possibly remove this restriction without needing a major api change.

Adding `priority` won't help with the locality weighted load balancing. e.g.,

```
C0{P0 us_east{H0}}
C1{P0 us_west{H1}}
```

Assuming the aggregate cluster supports merge endpoints by priority, 50% traffic will go to C0 and 50% traffic will go to C1.

I think the problem for aggregate cluster is that it cannot do a global load balancing once over all endpoints. It doesn't make sense sometimes because the aggregate cluster load balancer should respect the load balancer of the including clusters. However, if all clusters use same load balancing policy, should we provide a global load balancer instead of delegating the load balancing to the cluster? I think this will make aggregate cluster much more complex. WDYT? @snowp

Reply:
Everything you mentioned here makes sense. I am not at all familiar with how feasible it would be to having the priority - if it would never be added in the future then no reason to make the API more complex, but if it might happen in the future it may be useful?

Reply:
I added the `priority` in one commit and reverted afterwards because it's difficult to explain the behavior. I would prefer add that in the subsequent PR.

Reply:
This is all in `v2alpha` proto, so @yxue can always change the priority specification method later.

Reply:
sgtm